src/loader.rs

```diff
@@ -1,28 +1,162 @@
+pub fn find_shadowenv_paths(at: &Path) -> Result<Vec<PathBuf>, TraversalError> {
+    // First find the closest shadowenv.
+    let closest = match closest_shadowenv(at)? {
+        Some(closest) => closest,
+        None => return Ok(vec![]),
+    };
+
+    // Then find all parents recursively. Any validation errors bubble up.
+    let parents = resolve_shadowenv_parents(&closest)?;
+
+    Ok(iter::once(closest).chain(parents).collect())
+}
+
+/// Attempts to find the closest shadowenv folder to `at`.
+/// Returns a canonicalized path to the found shadowenv folder.
+fn closest_shadowenv(at: &Path) -> Result<Option<PathBuf>, TraversalError> {
+    for ancestor in at.ancestors() {
+        let dirpath = ancestor.join(SHADOWENV_DIR_NAME);
+        let metadata = match fs::metadata(&dirpath) {
+            Ok(metadata) => metadata,
+            Err(_) => continue, // Doens't exist or lacking permissions.
+        };
+
+        if metadata.is_dir() {
+            return Ok(Some(fs::canonicalize(dirpath)?));
         }
     }
-    Ok(roots)
+
+    Ok(None)
+}
+
+/// Recursively resolves parents.
+/// Expects `from_shadowenv` to be canonicalized.
+fn resolve_shadowenv_parents(from_shadowenv: &PathBuf) -> Result<Vec<PathBuf>, TraversalError> {
+    let parent_link = from_shadowenv.join(SHADOWENV_PARENT_LINK_NAME);
+
+    // `symlink_metadata`, opposed to `metadata`, doesn't resolve symlinks, which means that we can catch
+    // the case where symlink points to an non-existant file later on for a more precise error.
+    let metadata = match fs::symlink_metadata(&parent_link) {
+        Ok(metadata) => metadata,
+        // Symlink file itself doesn't exist or process is lacking access permissions.
+        Err(_) => return Ok(vec![]),
+    };
+
+    let previous_working_dir = env::current_dir().unwrap();
+    env::set_current_dir(from_shadowenv).unwrap();
+
+    // Must be a valid symlink.
+    // We need to resolve the symlink in context of the .shadowenv.d folder it is in.
+    let resolved_parent = fs::read_link(&parent_link)
+        .and_then(|resolved| resolved.canonicalize())
+        .map_err(|err| {
+            env::set_current_dir(&previous_working_dir).unwrap();
+
+            TraversalError::ResolveError {
+                parent_link_path: parent_link.to_string_lossy().to_string(),
+                error: if metadata.is_symlink() {
+                    err.to_string()
+                } else {
+                    "Not a symlink".to_owned()
+                },
+            }
+        })?;
+
+    // Restore working directory.
+    env::set_current_dir(previous_working_dir).unwrap();
+
+    // TODO: Refactor into better structure with the options.
+    let base_name = resolved_parent.file_name();
+    let base_name_stringified = base_name.map(|f| f.to_string_lossy());
+
+    // Must point to a SHADOWENV_DIR_NAME (e.g. `.shadowenv.d`).
+    if base_name_stringified != Some(Cow::Borrowed(SHADOWENV_DIR_NAME)) {
```

Comment:
Think this can be simplified to 

```suggestion
    if base_name_stringified.as_deref() != Some(SHADOWENV_DIR_NAME) {
```

Reply:
Good catch!